src/core/src/service_interfaces/TelemetryWriter.py

```diff
@@ -117,6 +118,76 @@ def get_disk_type(self):
+        features_keyvalue_list_str = os.getenv(Constants.AZURE_GUEST_AGENT_EXTENSION_SUPPORTED_FEATURES_ENV_VAR)
+        if features_keyvalue_list_str is None:
+            return False
+
+        features_keyvalue_list = json.loads(features_keyvalue_list_str)
+        return any(kv_pair for kv_pair in features_keyvalue_list if kv_pair['Key'] == Constants.TELEMETRY_EXTENSION_PIPELINE_SUPPORTED_KEY)
+
+    def get_agent_version(self):
+        """ Returns WALinuxAgent version, if installed. If not installed, returns None.
+            Returns a version string similar to: '2.2.49.2' """
+        cmd = "sudo waagent --version"
+        code, out = self.env_layer.run_command_output(cmd, False, False)
+        if code == 0:
+            ''' Command success, so the agent is installed and should return version info:
+                WALinuxAgent-2.2.49.2 running on sles 15.3
+                Python: 3.6.13
+                Goal state agent: 2.6.0.2 '''
+            return self.__agent_version_string_search(r'WALinuxAgent-\S+ running', out)
+
+        return None
+
+    def get_goal_state_agent_version(self):
+        """ Returns WALinuxAgent goal state agent version, if installed. If not installed, returns None.
+            Returns a version string similar to: '2.2.49.2' """
+        cmd = "sudo waagent --version"
+        code, out = self.env_layer.run_command_output(cmd, False, False)
+        if code == 0:
+            ''' Command success, so the agent is installed and should return version info:
+                WALinuxAgent-2.2.49.2 running on sles 15.3
+                Python: 3.6.13
+                Goal state agent: 2.6.0.2 '''
+            return self.__agent_version_string_search(r'Goal state agent: \S+', out)
+
+        return None
+
+    @staticmethod
+    def __agent_version_string_search(pattern, string):
+        """ Takes the output from waagent --version and extracts a specific agent version from it, if it exists. """
+        # Find substring containing the version
+        regex = re.compile(pattern)
+        version_str_search = regex.search(string)
+        if version_str_search is None:
+            return None
+
+        # Extract the version string
+        regex = re.compile(r'(\d+[.]*)+')
+        version_search = regex.search(version_str_search.group())
+        if version_search is None:
+            return None
+
+        return version_search.group()
+
+    def log_agent_information(self):
+        """ Logs WALinuxAgent version information. """
+        agent_version = self.get_agent_version()
+        if agent_version is None:
+            self.composite_logger.log('WALinuxAgent is not installed.', Constants.TelemetryEventLevel.Informational)
+        else:
+            self.composite_logger.log('WALinuxAgent version: {}\nGoal state agent version: {}'.format(agent_version, self.get_goal_state_agent_version()), Constants.TelemetryEventLevel.Informational)
+
```

Comment:
Code needs instrumentation. Is it possible to explain how the code executed and why it executed that way looking at a log file alone? 

For example, did the function at line 127 return False because the environment variable didn't exist or because the variable existed but the key didn't. On a related tangent, environment variables are case sensitive so take care around that.

Again for agent version search, did function fail because the agent returned nothing, or because it returned something that we weren't expecting to parse. Assume BVTs are all good and it's some customer coming and saying that our code didn't run and here's the log.

We must always be able to explain what code paths executed and why it executed that way from logs alone, else we'll be in this situation where we reactively need to push new code out and have a customer run the code again to understand what is actually happening. Which is a lot of overhead and will cause customer dissat.

Reply:
Thanks. Added some more logging for clarity so that the code path can be traced later.